rumqttd/src/segments/mod.rs

```diff
@@ -274,7 +276,7 @@ where
                 // debug!("start: {:?}, end: ({}, {}) done", orig_cursor, cursor.0, absolute_offset);
                 Ok(Position::Done {
                     start,
-                    end: (cursor.0, absolute_offset),
+                    end: (cursor.0 + 1, absolute_offset),
```

Comment:
This seems incorrect. Added failing case [here](https://github.com/bytebeamio/rumqtt/commit/2149d949a1f7ddfd6483a19a813ef2d99edd8762).

In test `active_segment_appends_and_reads_works` [we should get](https://github.com/bytebeamio/rumqtt/pull/593/files#diff-de9a4fd1a5e1abe08ac05206b8f03bc90d77d8d2ab40dc6d8fa903690d63628cR391-R392) `Done(0, 10)` instead of `Done(1, 10)`. This is because first segment in the CommitLog is still 90% empty. So our next read on the CommitLog should start from (0, 10). `cursor.0` should only increment when a segment is read fully. This is done [automatically](https://github.com/bytebeamio/rumqtt/pull/593/files#diff-de9a4fd1a5e1abe08ac05206b8f03bc90d77d8d2ab40dc6d8fa903690d63628cL199-L207) within `CommitLog::readv()`.

Reply:
good catch, from reading comments at https://github.com/bytebeamio/rumqtt/pull/593/files#diff-de9a4fd1a5e1abe08ac05206b8f03bc90d77d8d2ab40dc6d8fa903690d63628cL226-R236 i thought we only return `SegementPosition::Done` when we reach to the end of an segment. 

But when segment we are reading from is the active segment its not true because we return based on the `len` which is different from capacity of the segment. And when i saw `Done` inside `active_segment_appends_and_reads_works` I thought it was because of segment getting full and missed that segment has capacity to hold 100 elements not 10.

But the previous code is not correct as well because the invariant on `end` is that it is the cursor from where next read should begin and in few cases (see the test i added) it was breaking.

Reply:
fixed this by checking if current segment can hold more data or not.